Support building on Windows with Clang MSVC target

triplef opened the pull request:
Various changes to support building on Windows with Clang MSVC target and libobjc2.

The [GNUstep Windows MSVC Toolchain](https://github.com/gnustep/tools-windows-msvc) project has been using my branch with these changes for almost a year, but I’ve only now been able to confirm that the changes don’t cause any additional test failures on Linux.

triplef committed:
Add generated file Source/GNUmakefile to gitignore

triplef committed:
Update configure using autoconf.

triplef committed:
Add workaround for Clang bug on Windows MSVC

triplef committed:
Support building without pthread on Windows

Also fixed missing target detection.

stefanbidi commented on configure.ac:
```diff
@@ -77,6 +77,41 @@ AS_CASE([$gs_cv_c_data_model],
+# Check for pthread.h
+#--------------------------------------------------------------------
+AC_CHECK_HEADERS([pthread.h])
+if test $ac_cv_header_pthread_h = no ; then
```
Please use the "action-if-not-found" feature of AC_CHECK_HEADERS instead of explicitly checking for ac_cv_header_pthread_h. It's more portable and also easier to understand.

triplef replied:
I tried that, but I am getting the following warning when running `autoconf`. Do you have any suggestion how to fix that? I couldn’t figure it out from the link provided and searching Google...

```
configure.ac:9: warning: AC_REQUIRE: `AC_CONFIG_AUX_DIR_DEFAULT' was expanded before it was required
configure.ac:9: http://www.gnu.org/software/autoconf/manual/autoconf.html#Expanded-Before-Required
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
configure.ac:9: the top level
```

stefanbidi commented on configure.ac:
```diff
@@ -77,6 +77,41 @@ AS_CASE([$gs_cv_c_data_model],
+if test $ac_cv_header_pthread_h = no ; then
+  AC_MSG_ERROR([Unable to find pthread.h (needed for thread support).])
+fi
+AC_CHECK_LIB(pthread, pthread_join, pthread_ok=yes, pthread_ok=no)
```
I would prefer the use of AC_SEARCH_LIBS instead of AC_CHECK_LIB, here. Using it will allow you to skip the case statement below. Please double check, but I think the following should be enough:
`AC_SEARCH_LIBS([pthread pthreadG32 pthreadVC2], pthread_join, AC_DEFINE([HAVE_PTHREAD], ...))`

triplef replied:
Good point, thanks for the suggestion!

stefanbidi commented on Source/CFRunLoop.c:
```diff
@@ -430,6 +436,9 @@ CFRunLoopGetCurrent (void)
       pthread_setspecific(static_loopKey, rl);
     }
+#else
+  #warning CFRunLoopGetCurrent() not available (missing pthread library)
```
I would prefer this to be a configure check instead of a compile-time warning. Maybe adding a `--disable-cfrunloop` option to configure?

triplef replied:
I’ve added that configure option, and added the define `HAVE_CFRUNLOOP_SUPPORT` guarding the relevant code in CFRunLoop.c.

stefanbidi commented on Source/CFRunLoop.c:
```diff
@@ -389,6 +389,10 @@ GSRunLoopContextGet (CFRunLoopRef rl, CFStringRef mode)
 CFRunLoopCreate (void)
 {
+#ifdef _WIN32
+#warning CFRunLoop not implemented on Windows
```
Please consider using adding `--disable-cfrunloop` to configure.

stefanbidi commented on Source/CFStream.c:
```diff
@@ -428,14 +445,22 @@ CFWriteStreamFDCopyProperty (CFWriteStreamRef s, CFStringRef propertyName)
+      offset = SetFilePointer(stream->handle, 0, 0, FILE_CURRENT);
+      if (offset == INVALID_SET_FILE_POINTER)
+        {
+          CFWriteStreamSetError(s, GetLastError());
```
Is the output of `GetLastError()` compatible with the POSIX `errno` values?

triplef replied:
I think it returns Windows error codes, so probably not. Unfortunately the [documentation for `SetFilePointer()`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfilepointer) doesn’t specify the possible error codes so that we could convert them. It just says:
> If the function fails, the return value is INVALID_SET_FILE_POINTER. To get extended error information, call [GetLastError](https://docs.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror).

Not sure if there’s a better way to handle this but I’m open to suggestions.

stefanbidi commented on configure.ac:
```diff
@@ -6,6 +6,18 @@ AC_INIT([libgnustep-corebase], [0.2], [bug-gnustep@gnu.org])
 AC_CONFIG_SRCDIR([Source/CFBase.c])
 AC_CONFIG_HEADERS([Source/config.h])
 
+if test -n "$GNUSTEP_MAKEFILES"; then
```
Please use AS_IF() function instead of shell if-statement.

triplef replied:
I tried that, but I am getting the following warning when running `autoconf`. Do you have any suggestion how to fix that? I couldn’t figure it out from the link provided and searching Google...

```
configure.ac:9: warning: AC_REQUIRE: `AC_CONFIG_AUX_DIR_DEFAULT' was expanded before it was required
configure.ac:9: http://www.gnu.org/software/autoconf/manual/autoconf.html#Expanded-Before-Required
../../lib/m4sugar/m4sh.m4:639: AS_IF is expanded from...
configure.ac:9: the top level
```

stefanbidi commented on Source/CFString.c:
```diff
@@ -44,15 +44,20 @@
-#if HAVE_UNICODE_UNORM_H
+#if defined(HAVE_UNICODE_UNORM2_H)
+#include <unicode/unorm2.h>
+#elif defined(HAVE_UNICODE_UNORM_H)
```
If we're going to start using unorm2, I would like to remove support for the deprecated unorm function, then. UNormalizer2 has been available since version 4.4 and we required version >49. The only reason I used unorm was because this code dates back to version 4.0 of ICU.

stefanbidi commented on Source/CFString.c:
```diff
@@ -1641,6 +1646,25 @@ CFStringFold (CFMutableStringRef str, CFOptionFlags flags, CFLocaleRef locale)
     CFStringCaseMap (str, locale, flags, _kCFStringFold);
 }
 
+#ifdef __UNORM2_H__
```
Please remove support for unorm altogether. This should simplify the code significantly.

stefanbidi commented on configure.ac:
```diff
@@ -127,16 +127,34 @@ fi
-  [PKG_CHECK_MODULES([ICU], [icu-i18n > 49.0])
-  AC_CHECK_HEADERS([unicode/ucal.h unicode/uchar.h unicode/ucnv.h unicode/ucol.h unicode/ucurr.h unicode/udat.h unicode/udatpg.h unicode/uloc.h unicode/ulocdata.h unicode/unorm.h unicode/unum.h unicode/usearch.h unicode/ustring.h unicode/utrans.h],
-  [], AC_MSG_ERROR([Could not find required ICU headers.]))
+AS_IF([test x"$with_icu" != xno], [
```
I feel there should be a way to simplify this check. We really only use PKG_CHECK_MODULES to check the version of the installed ICU library. I don't really know how ICU is packaged in Windows, but it sounds like only icu.h is provided. Maybe this check for icu.h can be done directly in AC_CHECK_HEADERS's "action-if-not-found" parameter.

triplef replied:
I’ve tried to simplify it as much as I could. It’s a bit complex because on Windows we need to try using the bundled ICU first in order to avoid running PKG_CHECK_MODULES and getting an error, but we want to fall back to pkg-config if the bundled ICU is not found (on older Windows releases). Please let me know if you have any suggestions on how to simplify this further.

stefanbidi reviewed (changes requested):
There are a few areas of concern in your PR. I've added comments and hope they can be addressed.

triplef commented:
Thank you for your thorough review! I’ll try to integrate the requested changes, all makes sense.

triplef committed:
Add support for building without pthread library.

triplef committed:
Fixed support for building on Windows.

CFRunLoop and directory enumeration in CFURLAccess is currently not implemented on Windows, and Windows-specific implementations of CFSocket, CFStream, and CFURLAccess are largely untested so far.

triplef committed:
Add support for using ICU DLL from Windows 10

Requires Windows 10 version 1903 (May 2019 Update):
https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-

triplef committed:
Add configure option to disable CFRunLoop support

triplef committed:
Fix logic in CFGregorianDateIsValid()

triplef committed:
Use unorm2 API instead of deprecated unorm

UNormalizer2 has been available since version 4.4 and we require ICU version > 49.0.

triplef committed:
Add ICU_CFLAGS to INCLUDE_FLAGS

Fixes finding ICU headers if they are not in existing includes.

triplef commented:
I’ve incorporated most of your feedback, please see the two remaining issues above.

triplef commented:
While working on this I also found a logic issue in `CFGregorianDateIsValid()` and fixed the ICU header search when building for Android.
